provider/clickhouse.go

```diff
@@ -0,0 +1,1305 @@
+	query, err := store.createsqlPrimaryTableQuery(name, schema)
+	if err != nil {
+		return nil, err
+	}
+	_, err = db.Exec(query)
+	if err != nil {
+		return nil, err
+	}
+	return &clickhousePrimaryTable{
+		db:     db,
+		name:   name,
+		schema: schema,
+		query:  store.query,
+	}, nil
+}
+
+func (store *clickHouseOfflineStore) getValueIndex(columns []TableColumn) int {
+	for i, column := range columns {
+		if column.Name == "value" {
+			return i
+		}
+	}
+	return -1
+}
+
+func (store *clickHouseOfflineStore) newsqlOfflineTable(db *sql.DB, name string, valueType ValueType) (*clickhouseOfflineTable, error) {
+	columnType, err := determineColumnType(valueType)
+	if err != nil {
+		return nil, fmt.Errorf("could not determine column type: %v", err)
+	}
+	tableCreateQry := store.query.newSQLOfflineTable(name, columnType)
+	_, err = db.Exec(tableCreateQry)
+	if err != nil {
+		return nil, fmt.Errorf("could not create table query: %v", err)
+	}
+	return &clickhouseOfflineTable{
+		db:    db,
+		name:  name,
+		query: store.query,
+	}, nil
+}
+
+func (store *clickHouseOfflineStore) getsqlResourceTable(id ResourceID) (*clickhouseOfflineTable, error) {
+	if exists, err := store.tableExists(id); err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, &TableNotFound{id.Name, id.Variant}
+	}
+	table, err := store.getResourceTableName(id)
+	if err != nil {
+		return nil, err
+	}
+	return &clickhouseOfflineTable{
+		db:    store.db,
+		name:  table,
+		query: store.query,
+	}, nil
+}
+
+func (store *clickHouseOfflineStore) materializationExists(id MaterializationID) (bool, error) {
+	tableName := store.getMaterializationTableName(id)
+	getMatQry := store.query.materializationExists()
+	n := -1
+	err := store.db.QueryRow(getMatQry, tableName).Scan(&n)
+	if err != nil {
+		return false, err
+	}
+	if n == 0 {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (store *clickHouseOfflineStore) getMaterializationTableName(id MaterializationID) string {
+	return fmt.Sprintf("featureform_materialization_%s", id)
+}
+
+func clickhouseOfflineStoreFactory(config pc.SerializedConfig) (Provider, error) {
+	store, err := NewClickHouseOfflineStore(config)
+	if err != nil {
+		return nil, err
+	}
+	return store, nil
+}
+
+func NewClickHouseOfflineStore(config pc.SerializedConfig) (*clickHouseOfflineStore, error) {
+	cc := pc.ClickHouseConfig{}
+	if err := cc.Deserialize(config); err != nil {
+		return nil, NewProviderError(Runtime, pt.ClickHouseOffline, ConfigDeserialize, err.Error())
+	}
+	var t *tls.Config
+	if cc.SSL {
+		t = &tls.Config{}
+	}
+	db := clickhouse.OpenDB(&clickhouse.Options{
+		Addr: []string{fmt.Sprintf("%s:%d", cc.Host, cc.Port)},
+		Auth: clickhouse.Auth{
+			Database: cc.Database,
+			Username: cc.Username,
+			Password: cc.Password,
+		},
+		Settings: clickhouse.Settings{
+			"final": "1",
+		},
+		TLS: t,
+	})
+	queries := clickhouseSQLQueries{}
+	// numeric and should work
+	queries.setVariableBinding(PostgresBindingStyle)
```

Comment:
just want to make sure that this is the right value?

Reply:
Yes this works